engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,167 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	commenced          bool // if stopping process has started. We disallow any changes then.
+	set                bool // whether stop at height has been set at all, it its envisioned most of the time it won't
+	paused             bool // if execution is paused
+	stopAfterExecuting flow.Identifier
+	log                zerolog.Logger
+}
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
+	return &StopControl{
+		paused:    paused,
+		commenced: paused,
+		log:       log,
+	}
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.paused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (bool, uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldSet := s.set
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.commenced {
+		return oldSet, oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping already in progress for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	s.set = true
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldSet, oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - set, whether values are set
+//   - height
+//   - crash
+func (s *StopControl) GetStopHeight() (bool, uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.set, s.height, s.crash
+}
+
+// BlockProcessable should be called when new block is processable.
+// It returns boolean indicating if processing should skip this block.
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if !s.set {
+		return false
+	}
+
+	if s.paused {
+		return true
+	}
```

Comment:
we should handle paused first, if it's true, then we must skip. if is false, then we check set

```suggestion
	if s.paused {
		return true
	}
	
	if !s.set {
		return false
	}


```

Reply:
if `set` is false, `pause` value doesn't matter (as it has not been explicitly set)

Reply:
The idea is that `set` determines if this whole module is even active. Maybe it would be better to rename this to `active` or something along those lines. 
Maybe it would avoid confusion if I turn this into a proper state machine - lifecycle like 
`off` -> `set` -> `skipped` -> `paused` -> `stopped`  should be cleaner (not sure if thats fully cycle here, just an example).